### ISSUE TITLE
Inputs added by input relays are hard to decode

### ISSUE THREAD

**Issue by guidanoli** [cartesi/rollups#134]
## 📚 Context

Let us start with a short introduction on inputs. In order to interact with a DApp in a verifiable way, one must send an input to the `InputBox` contract. Off-chain, the node eventually picks up this new input by listening to `InputAdded` events emitted by this contract, and forwards it to the Cartesi Machine running the DApp back-end.

Whenever a DApp receives an input, it generally follows these steps: check if the input sender and/or the input format are sound; then apply the effects of accepting the input. In some cases, the DApp only checks whether the input comes from a trusted smart contract and assumes the input format is correct. In other cases, the DApp assumes the input comes from an EOA (externally-owned account) and only checks if the input is properly formatted. For example, it may check whether the input is a valid JSON string, and implements a given schema. 

Input relays are smart contracts that relay information pertaining to the base layer to applications running on the execution layer through inputs. Portals, for example, relay information about asset transfers. Given that input relays are meant to be used by all applications, **we should make an effort to make the inputs added by them simple to validate and decode**.

Currently, the input relays encode inputs with Solidity's [non-standard packed mode](https://docs.soliditylang.org/en/v0.8.19/abi-spec.html#non-standard-packed-mode). Compared to the [strict encoding mode](https://docs.soliditylang.org/en/v0.8.19/abi-spec.html#strict-encoding-mode), which is ubiquitously used by Solidity, the packed mode yields smaller binary outputs, which is interesting from a gas consumption point-of-view. The problem with this mode is that, for some schemas, the encoding is not injective, meaning there is no associated decode function.

However, even though the input relays use schemas for which there are associated decode functions, this decoding algorithm is not generally defined by libraries that manipulate data encoded according to the Solidity ABI. Instead, they generally only define encode/decode entry points for the strict encoding mode. The DApp developer, therefore, has to either implement the decoding algorithm themselves, which may contain vulnerabilities that can be exploited by bad actors.

## ✔️ Solution

To promote a smoother developer experience with code reusability, and to avoid ad-hoc, error-prone implementations of important algorithms, we would like to switch the encoding of inputs added by input relays from the non-standard packed mode to the strict encoding mode.

## :people_holding_hands: Acknowledgements

Thanks, @fmoura @miltonjonat @diegonehab @GCdePaula, for pointing out this problem to me.

**Comment by miltonjonat** [comment]
Hi there @guidanoli. To be honest, I think this is also relevant for the inputs added by the Portals. Using standard (non-packed) encoding for all inputs would help a lot, ensuring that we can easily specify encoders and decoders for any input.
That being said, I do hope that most of this complexity will be hidden away by using higher-level frameworks, which will take care of this decoding/encoding.
@tuler may have interesting comments on this too

**Comment by guidanoli** [comment]
Hi @miltonjonat, sorry for the lack of clarity of my part.
By "input relay", I'm also including portals. See #33.

**Comment by augustoteixeira** [comment]
I agree that we should be using Standard Encoding for "relayed inputs", but I think it is good to make an observation about "non-relayed inputs" as well (those from EOA's).

The cost of adding external inputs is usually dictated by the L1 cost of data availability, so it may not be a good solution to use the Standard Ethereum Encoding (nor JSON for that matter) because they have bad compression characteristics.

In that direction, we could explore other encodings, such as https://github.com/jamesmunns/postcard
